CRM/Core/Resources.php

```diff
@@ -809,4 +810,25 @@ public static function getEntityRefFilters() {
     return $filters;
   }
 
+  /**
+   * @param string $url
+   * @return string
+   */
+  public function addCacheCode($url) {
+    parse_str(parse_url($url, PHP_URL_QUERY), $queryParts);
```

Comment:
At first I was a little nervous about this about adding this to `getUrl()` -- my gut says to expect ballpark of ~50 calls to `getUrl()` in a given page-request. (Maybe the real number is 4. Maybe it's 100. It's probably not 1,000,000.)  So I did a quick benchmark comparing these three techniques for checking for `?r=`:

```
  parse_str(parse_url($url, PHP_URL_QUERY), $queryParts);
  $existing = isset($queryParts['r']) ? $queryParts['r'] : null;

  $existing = (strpos($url, '?r=') !== FALSE) || (strpos($url, '&r=') !== FALSE);

  $existing = preg_match(';[\&\?]r=;', $url);
```

`strpos()` takes about half as much time as `parse_str(parse_url())`, but I had to really ramp up the scale to make it observable:
 * If processing 2,000 URLs in one request, the difference is consistently 10ms vs 5ms on my Macbook Pro.
 * At a more realistic volume of 100 URLs, the difference is 0.3ms vs 0.6ms.

So I think it's officially negligible at likely scale.